### DISCUSSION TITLE
Adjustment for the Autoscaling-Flag of Eclipse SWT

### DISCUSSION THREAD

**Discussion by Michael5601** [eclipse-platform/eclipse.platform.swt#1693] [General]
The proposed SVG rasterization feature in this [PR](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1638) introduces the ability to rasterize SVGs at arbitrary resolutions.

As this feature is an enhancement of the current API the output resolution of the rasterized image is selected by calling it with a specific zoom factor. This zoom factor is determined by SWT according to the flag `Dswt.autoScale`. 

Right now the standard value for this flag is `integer200` which works as follows:
- For native zoom levels on the main monitor between 100% and 174%, a zoom factor of 100% is applied.
- For zoom levels between 175% and 200%, a zoom factor of 200% is applied.

Under this default setting, no image scaling is performed. This behavior is intentional for PNGs, as scaling them mostly produces poor results. However, for SVGs, this approach means the correct resolution is never provided.

To fully leverage the benefits of the SVG rasterization feature, the `Dswt.autoScale`-flag must be set to `exact`, enabling the generation of images at the correct resolution.

In summary this discussion is about the question which value should be chosen as standard for the `Dswt.autoScale`-flag. 

My suggestion is as follows:
Since there is no direct way to notify users about the new SVG feature, such as through `@Deprecated` annotations, I propose setting the default value of `Dswt.autoScale` to `exact`. Users can still modify this setting in the `eclipse.ini`-file, as they currently do to address issues with small icons on high-DPI monitors.

**Comment by BeckerWdf** [comment]
> I propose setting the default value of `Dswt.autoScale` to `exact`

What will be the effect on existing PNGs? Will scaling them with `exact` create poor results?

**Reply by Michael5601** [reply]
Yes unfortunately it will be the case if no PNGs for the target resolution are available. Only in this case the scaling will be performed which leads to poor results. 

On the other hand the current solution is also poor in my opinion as all icons are really small which leads to users manually activating the scaling and accepting the bad scaling results.

**Reply by BeckerWdf** [reply]
> Yes unfortunately it will be the case if no PNGs for the target resolution are available. Only in this case the scaling will be performed which leads to poor results.
> 
> On the other hand the current solution is also poor in my opinion as all icons are really small which leads to users manually activating the scaling and accepting the bad scaling results.

That's bad. We do the SVG stuff so that Eclipse looks nicer. And now this would make it look more ugly.
Can't we solve this in a different way. So that only the SVG scaling is affected?

**Reply by Michael5601** [reply]
We had a discussion about a separate flag for SVG scaling that is not affected by the `Dswt.autoScale`-flag. 

But the problem with this is that in the worst case some icons will be small (PNG) and some will be in the correct size (SVG). Thus I would not recommend this solution.

**Reply by HeikoKlare** [reply]
> That's bad. We do the SVG stuff so that Eclipse looks nicer. And now this would make it look more ugly.
> Can't we solve this in a different way. So that only the SVG scaling is affected?

What would be the expected result of this? Should SVG-based icons be scaled and PNG-based icons be unscaled? Then you would result in a UI with differently scaled icons. You will probably have worst results at 150%, where you have SVG-based icons at 150% and PNG-based icons at 100%. Of course, scaled PNG-based icons do not look nice, but once we have SVG support they should soon make up a minority of the icons used.

I have to admit that I never really understood the existing solution/default of not scaling icons. In my opinion, having 100% icons at 150% UI scale is a bad experience. In addition, the "correct" scaled value (100% when the native is 150%) is also used at other places and not only for icons, which makes the code complicated and the UI buggy. Thus, I would be in favor of getting rid of the disctinction between "native zoom" and "device zoom" in general.

We just need to keep in mind that, as mentioned above, the `swt.autoScale` flag does not only affect icons but UI scaling in general. E.g., it also affects sizes of controls. This usually has no affect, as sized are computed based on fonts (which always use native zoom) and/or using a layout manager. In those cases, where none of the two applies and a custom size is set for a control, there will be a difference when the `swt.autoScale` flag is changed. I would at least views that do not use a layout manager, in particular diagrams based on GEF, be affected by this.
@azoitl do you have any experience with changing the `swt.autoScale` flag to `quarter` or `exact` when using GEF Classic?

**Reply by BeckerWdf** [reply]
> What would be the expected result of this? Should SVG-based icons be scaled and PNG-based icons be unscaled? Then you would result in a UI with differently scaled icons. You will probably have worst results at 150%, where you have SVG-based icons at 150% and PNG-based icons at 100%. Of course, scaled PNG-based icons do not look nice, but once we have SVG support they should soon make up a minority of the icons used.

You are right. That might be worse. Maybe we need to see in screenshot "how" bad the PNGs would look like.

> Of course, scaled PNG-based icons do not look nice, but once we have SVG support they should soon make up a minority of the icons used.

That's true. So if user's complain that icons look bad the fix would to be replace them with SVGs.

**Reply by HeikoKlare** [reply]
> Maybe we need to see in screenshot "how" bad the PNGs would look like.

This is how an SDK looks at 150% (i.e., with scaled icons) with `swt.autoScale=quarter`. The icons are not great, but I still prefer them over being wrongly sized. Note that you need to open the image at original size to properly see how the icons look like.

![image](https://github.com/user-attachments/assets/f0382238-adb9-4819-8d94-647fa3dc5a5b)

**Reply by BeckerWdf** [reply]
And how does it look like with today's settings? I would like to compare them next to each other.

**Reply by HeikoKlare** [reply]
This is how it looks with `swt.autoScale=integer200`, i.e., with today's default:
![image](https://github.com/user-attachments/assets/80ea02ad-9d5e-4e64-b680-01a3f865d627)

Here is a side-by-side comparison (left is `integer200`/current, right is `quarter`):
![image](https://github.com/user-attachments/assets/74b79735-7d7d-49e5-b87e-d2f60ee81bd0)

All screenshots are taken with 150% scaling.

**Reply by azoitl** [reply]
@HeikoKlare unfortunately I have not yet experience with the auto scaling flag. But I know that at least in the RCP using GEF Classic that I'm developing I have issues with certain elements of my GEF drawings not being correctly scaled. This is something I have on my to do list for the beginning of 2025. 

I had some discussion with @ptziegler at OCX25. I had the impression he has already more experience then me.

**Reply by azoitl** [reply]
After re-investigating our drawing issues in Eclipse 4diac IDE's GEF editors and the discussion here I finally understood what is the problem for any GEF based editor where fractional native scaling is set in the OS and Eclipse Platform is used with the default `swt.autoScaling` option: 

    - any text based drawings are scaled with with autoscaling exact because of the OS based text rendering
    - any graphics only not based on font sizes are not scaled at all. 

This can lead to strange looking diagrams. I've tested the `-Dswt.autoScale=exact` option and this produces much more reliable outputs.  For 4diac IDE we will use for now per default exact. 

For comparison here a compilation of screenshots with first line default  `-Dswt.autoScale` settings and the second line with `exact`, done on Windows.

![scaling_1](https://github.com/user-attachments/assets/46f573f7-0e6f-4449-b405-02e589b23b67)

One thing that I found interesting is that for 175% native scaling the drawings get very blury. Not yet sure what is the reason for that.

**Reply by HeikoKlare** [reply]
Thank you for the investigation! I agree that the results with scaling set to `exact` look more consistent across the different scale levels. This is the result of images and control bounds being sized "properly". It should produce the same results with `quarter`.

The blurry scaling on 175% is interesting and unexpected. It looks as if the texts have been rescaled after rendering, but I don't see why that should happen as I would expect text to be rendered in the exact scaling required for presentation in the UI.

**Reply by azoitl** [reply]
> Thank you for the investigation! I agree that the results with scaling set to `exact` look more consistent across the different scale levels. This is the result of images and control bounds being sized "properly". 

I would emphasize that this is a bit heaver then only images and bounds. This is for anything that is drawn in GEF based on pixels. In the example above it is the drop shadow, the green triangles, etc. One element for which it is extremely bad is the grid. You can not see that in the example above because in 4diac IDE we reworked the Grid a month ago. Now our grid is based on the distance between two pins (two of the triangles). As this is based on the font size and font size is handled by the OS this is now better. But for any GEF Classic base editor the impact is bigger then I had expected. (I forgot to tag @ptziegler in my last post: @ptziegler this may also be interesting for you). Also we have to keep this in mind for SWT 31.

> It should produce the same results with `quarter`.

Yes I would expect that `quater` is for the native scaling settings that the test machine offered generate the same results. 

> The blurry scaling on 175% is interesting and unexpected. It looks as if the texts have been rescaled after rendering, but I don't see why that should happen as I would expect text to be rendered in the exact scaling required for presentation in the UI.

That is the reason I wanted to point it out. What I can say that it was produced on the same machine as the others with the same setup (except native scaling) with the same GEF zoom factor. Just to rule a view things out.

**Comment by BeckerWdf** [comment]
ok from the screenshots I would say that this is not too bad

**Reply by HannesWell** [reply]
I agree that changing the default of `swt.autoScale` to `exact` or maybe `quarter` (for a grace period) seems the best compromise.
Scaling just SVG will look far worse than having some PNGs that are only poorly scaled.
But we should only change this default have most of the icons we own in the SDK are migrated to SVG.
This will also make users and developers aware of the new support SVG rendering (of course we should still announce it).

What I generally miss with regard to scaling is the ability to adjust the scaling in the preference UI, even if it would just write the corresponding property value to the ini file.

**Reply by HeikoKlare** [reply]
I agree that it would sometimes be nice to define the autoscale value via a preference, but we need to be aware that we have far too many scaling modes available and they are not all working as expected (whatever may be expected at all). I guess not so many users know about the ability to change them. If we now provided UI for changing them, the expectations regarding keeping, maintaining and improving them could increase, while I would be in favor of just getting rid of all them. In my opinion, we should aim for only having "exact" (and maybe a custom definition of the zoom if anyone required that) but no "integer", "integer200", "half" or "quarter" as they are all just workaround for issues that should be fixed themselves (or I have been completely missing some specific need for them).

For the new "monitor-specific scaling" that we are currently introducing for Windows, we decided to only support "quarter", "exact" and explicit zoom values (including "false"), as the currently used "integer"/"integer200" (and also "half") modes are too complicated to implement without any real need (except for sharp image scaling, which will mostly be resolved by SVGs).
We will thus make "monitor-specific scaling" use "quarter" autoscale mode (if no other mode else is specified) or any other of the accepted modes in case they are explicitly specified: 
- https://github.com/eclipse-platform/eclipse.platform.swt/pull/1721

This may also help us in the sense that it pushes people to using such an autoscale mode and that it allows us to more easily test with "quarter" mode, as the monitor-specific scaling can be enabled via the preference UI.

**Reply by HannesWell** [reply]
> I agree that it would sometimes be nice to define the autoscale value via a preference, but we need to be aware that we have far too many scaling modes available and they are not all working as expected (whatever may be expected at all).

Unfortunately yes. A colleague of mine uses a 4K monitor under Linux for quite some time now hand had to find a non-trivial combination of env-variables to make it somewhat usable but still has multiple oddities.  



> In my opinion, we should aim for only having "exact" (and maybe a custom definition of the zoom if anyone required that) but no "integer", "integer200", "half" or "quarter" as they are all just workaround for issues that should be fixed themselves (or I have been completely missing some specific need for them).

Yes, absolutely. Ideally Eclipse would just look good on a high-DPI monitor out of the box respecting the scaling defined by the OS.

But if we change the default of `swt.autoScale` in SWT, we also have to adapt the Equinox-launcher, otherwise the splash-screen looks bad:
- https://github.com/eclipse-equinox/equinox/blob/a1c46e90ca59dc3e15dde1be7df95fb3c1da11a4/features/org.eclipse.equinox.executable.feature/library/win32/eclipseWin.c#L705-L733

**Reply by Michael5601** [reply]
> But if we change the default of `swt.autoScale` in SWT, we also have to adapt the Equinox-launcher, otherwise the splash-screen looks bad:

I guess this won't be the only place where we need to change the behaviour regarding the flag. As @HeikoKlare said there are multiple places that need adjustment. Another example is the `CompositeImageDescriptor` in Platform UI that has a method `supportsZoomLevel` that only returns true when the scaling factor is a multiple of 100% (e.g. 100%, 200%, ...). Even if the flag is set to `exact` by default composite icons would be rendered incorrectly for scaling factors like 125%. 

https://github.com/eclipse-platform/eclipse.platform.ui/blob/9b70eea07aba502bcdac733304560fe275d27cb3/bundles/org.eclipse.jface/src/org/eclipse/jface/resource/CompositeImageDescriptor.java#L449

In the proposed SVG feature I changed this method to always return true but this is not a  good solution.

**Reply by HeikoKlare** [reply]
> Another example is the `CompositeImageDescriptor` in Platform UI that has a method `supportsZoomLevel` that only returns true when the scaling factor is a multiple of 100% (e.g. 100%, 200%, ...). Even if the flag is set to `exact` by default composite icons would be rendered incorrectly for scaling factors like 125%.

I am currently on that method as I found that the issues related to this specific behavior (as well as the reasons for using the bad scaling of images/icons in general) are already resolved, so we may easily get rid of the bad icon scaling as well as this specific handling in `CompositeImageDescriptor`. I am currently testing according changes and will provide PRs soon.
